src/Storages/StorageS3Distributed.h

```diff
@@ -0,0 +1,157 @@
+#pragma once
+
+#include <Common/config.h>
+
+#if USE_AWS_S3
+
+#include "Client/Connection.h"
+#include "Interpreters/Cluster.h"
+#include "Storages/IStorage.h"
+#include "Storages/StorageS3.h"
+
+#include <memory>
+#include <optional>
+#include "ext/shared_ptr_helper.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+
+class Context;
+
+struct ClientAuthentificationBuilder
+{
+    String access_key_id;
+    String secret_access_key;
+    UInt64 max_connections;
+};
+
+using QueryId = std::string;
+using Task = std::string;
+using Tasks = std::vector<Task>;
+using TasksIterator = Tasks::iterator;
+
+class S3NextTaskResolver
+{
+public:
+    S3NextTaskResolver(QueryId query_id, Tasks && all_tasks)
+        : id(query_id)
+        , tasks(all_tasks)
+        , current(tasks.begin())
+    {}
+
+    std::string next()
+    {
+        auto it = current;
+        ++current;
+        return it == tasks.end() ? "" : *it;
+    }
+
+    std::string getId()
+    {
+        return id;
+    }
+
+private:
+    QueryId id;
+    Tasks tasks;
+    TasksIterator current;
+};
+
+using S3NextTaskResolverPtr = std::shared_ptr<S3NextTaskResolver>;
+
+class TaskSupervisor
+{
+public:
+    using QueryId = std::string;
+
+    TaskSupervisor() = default;
+
+    static TaskSupervisor & instance()
+    {
+        static TaskSupervisor task_manager;
+        return task_manager;
+    }
+
+    void registerNextTaskResolver(S3NextTaskResolverPtr resolver)
+    {
+        std::lock_guard lock(mutex);
+        auto & target = dict[resolver->getId()];
+        if (target)
+            throw Exception(fmt::format("NextTaskResolver with name {} is already registered for query {}",
+                target->getId(), resolver->getId()), ErrorCodes::LOGICAL_ERROR);
+        target = std::move(resolver);
+    }
+
+
+    Task getNextTaskForId(const QueryId & id)
+    {
+        std::lock_guard lock(mutex);
+        auto it = dict.find(id);
+        if (it == dict.end())
+            return "";
+        auto answer = it->second->next();
+        if (answer.empty())
+            dict.erase(it); 
+        return answer;
+    }
+
+private:
+    using ResolverDict = std::unordered_map<QueryId, S3NextTaskResolverPtr>;
+    ResolverDict dict;
+    std::mutex mutex;
+};
+
+class StorageS3Distributed : public ext::shared_ptr_helper<StorageS3Distributed>, public IStorage 
+{
+    friend struct ext::shared_ptr_helper<StorageS3Distributed>;
+public:
+    std::string getName() const override { return "S3Distributed"; }
+
+    Pipe read(
```

Comment:
Note that `StorageDistributed` also support rename and alter.
(In case we want to add `S3Distributed` engine, not just table function).

Reply:
Also `getQueryProcessingStage` is not implemented, so it is always `FetchColumns`.
Aggregation on remote replicas does not work

Reply:
> Also `getQueryProcessingStage` is not implemented, so it is always `FetchColumns`.
> Aggregation on remote replicas does not work

Fixed